src/envoy/http/mixer/filter.cc

```diff
@@ -30,6 +31,49 @@ namespace Envoy {
 namespace Http {
 namespace Mixer {
 
+namespace {
+
+// detect gRPC message boundaries: each message is prefixed by 5 bytes
+// length-prefix https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md
+void incrementCounter(Buffer::Instance& data, GrpcMessageCounter* counter) {
+  uint64_t pos = 0;
```

Comment:
I am not sure if it is correct to add pos=0 here. 
Is Filter::encodeData() called only once in the response pipeline, or called multiple times until all response data are sent out?
If Filter::encodeData could be invoked multiple times, then in one call, Buffer::Instance& data might have only 1 byte Compressed-Flag + 1~3 bytes Message-Length, in the next call, Buffer::Instance& data might have the rest bytes of Message-Length. The worse case could be that Buffer::Instance& data only has one byte in a call.

Reply:
Counting messages in this way might slow down the gRPC streaming. I am not sure how much difference the performance would become. Maybe we need to conduct some load test for analysis?

Reply:
I'm not sure what the concern here. The state machine (ExpectByte0, etc) works across data buffer boundaries. The pos variable is the index into the current data buffer bytes. The two work independently.

Reply:
Regarding the perf impact: yes, it is important to measure that. My feeling is that relative to what the rest of envoy is doing, it is probably not as big as it might seem.

Reply:
agreed it shouldn't be big as it only process very few pieces of the data.

Reply:
> I'm not sure what the concern here. The state machine (ExpectByte0, etc) works across data buffer boundaries. The pos variable is the index into the current data buffer bytes. The two work independently.

I am thinking about a scenario like this. Assume Filter::encodeData(Buffer::Instance& data, bool) is called twice, in the first call, the passed Buffer::Instance& data has three bytes, 1 byte of Compressed-Flag and 2 bytes of Message-Length. In the second call, the passed Buffer::Instance& data has 2 bytes of Message-Length and Message. I am not sure if this will happen, so I would like to confirm whether we need to handle that case.

Reply:
What happens is that on the second data buffer, we have state `ExpectByte3` and `pos = 0`. Then after first two bytes in the second buffer, the state is `ExpectMessage` and the rest of the buffer is skipped over up to the consumed length prefix.

Reply:
I see, the state is kept in counter->state.